vmdb/spec/requests/shared_examples/api.rb

```diff
@@ -0,0 +1,228 @@
+    expect(hash).to have_key(key)
+    expect(hash[key]).to match(value)
+  end
+end
+
+def call_custom_expects(method)
+  expect(respond_to?(method)).to be_true
+  public_send(method)
+end
+
+shared_examples_for "bad_request" do |error_message|
+  it "request failed with bad request (400) #{error_message}" do
+    expect(@success).to be_false
+    expect(@code).to eq(400)
+
+    if error_message.present?
+      expect(@result).to have_key("error")
+      expect(@result["error"]["message"]).to match(error_message)
+    end
+  end
+end
+
+shared_examples_for "user_unauthorized" do
+  it "request failed with unauthorized (401)" do
+    expect(@success).to be_false
+    expect(@code).to eq(401)
+  end
+end
+
+shared_examples_for "request_forbidden" do
+  it "request failed with request forbidden (403)" do
+    expect(@success).to be_false
+    expect(@code).to eq(403)
+  end
+end
+
+shared_examples_for "resource_not_found" do
+  it "request failed with resource not found (404)" do
+    expect(@success).to be_false
+    expect(@code).to eq(404)
+  end
+end
+
+shared_examples_for "request_success" do |options|
+  it "request succeeded with (200)" do
+    options ||= {}
+    expect(@success).to be_true
+    expect(@code).to eq(200)
+    expect_result_to_include_data(*options[:includes_data]) if options[:includes_data].present?
+    call_custom_expects(options[:custom_expects]) if options[:custom_expects].present?
+  end
+end
+
+shared_examples_for "request_success_no_content" do
+  it "request succeeded with no content (204)" do
+    expect(@success).to be_true
+    expect(@code).to eq(204)
+  end
+end
+
+shared_examples_for "empty_query_result" do |collection|
+  it "collection result for #{collection} to be empty" do
+    expect(@success).to be_true
+    expect(@code).to eq(200)
+    expect(@result).to have_key("name")
+    expect(@result["name"]).to eq(collection.to_s)
+    expect(@result["resources"]).to be_empty
+  end
+end
+
+shared_examples_for "query_result" do |collection, size, options|
+  it "collection result for #{collection} to be empty" do
+    options ||= {}
+    expect(@success).to be_true
+    expect(@code).to eq(200)
+    expect(@result).to have_key("name")
+    expect(@result["name"]).to eq(collection.to_s)
+    expect(@result["count"]).to eq(fetch_value(options["count"])) if options["count"]
+    expect(@result["subcount"]).to eq(fetch_value(size))
+    expect(@result["resources"].size).to eq(fetch_value(size))
+    expect_result_to_include_hrefs(*options[:includes_hrefs]) if options[:includes_hrefs].present?
+    expect_result_to_include_data(*options[:includes_data]) if options[:includes_data].present?
+  end
+end
+
+shared_examples_for "results_include_hrefs" do |collection, href_list|
+  it "result resources include hrefs" do
+    expect_result_to_include_hrefs(collection, href_list)
+  end
+end
+
+shared_examples_for "results_include_data" do |collection, data|
+  it "result resources include data" do
```

Comment:
I can go either way, going the shared examples, might be better longer term once we have 3.2.  The only negative vs custom matchers, the output is a bit more verbose shown here:

it/custom matcher way:

```
ApiController
  Vm start action
    starts a powered on vm
    starts an invalid vm without appropriate role
    starts a vm
    starts an invalid vm
    starts multiple vms 
```

context/shared_examples way:

```
ApiController
  Vm start action
    starts a powered on vm
      processed a single action
    starts an invalid vm without appropriate role
      request failed with request forbidden (403)
    starts a vm
      processed a single action
    starts multiple vms 
      processed multiple actions
    starts an invalid vm
      request failed with resource not found (404)
```

 but once in rspec 3.2 and we add multiple shared examples in complex responses,  failures could be more specific.

Again, I can go either way. @kbrock @fryguy what are your preferences ?  going to custom matchers will take some work so I don't want to have to change back to this style later.  Thanks.